Sources/SourceKitLSP/Swift/SemanticTokensManager.swift

```diff
@@ -0,0 +1,90 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LanguageServerProtocol
+
+/// Keeps track of the semantic tokens that sourcekitd has sent us for given
+/// document snapshots.
+actor SemanticTokensManager {
+  private var semanticTokens: [DocumentSnapshot.ID: [SyntaxHighlightingToken]] = [:]
+
+  /// The semantic tokens for the given snapshot or `nil` if no semantic tokens
+  /// have been computed yet.
+  func semanticTokens(for snapshotID: DocumentSnapshot.ID) -> [SyntaxHighlightingToken]? {
```

Comment:
Personally I would prefer to wrap `[SyntaxHighlightingToken]` in its own struct, e.g then you could put members like `lspEncoded` and `mergingTokens` on it, and maybe move the edit processing logic onto it too. I don't feel that strongly about it though.

Reply:
Good idea. It’s a slightly bigger change and I’ll do it in a follow-up PR.